Idempotenz issue

josojo opened the pull request.

josojo commented:
Currently it is only 92 branch coverage. I will add tests on Monday

fleupold commented on contracts/DxMgnPool.sol:
```diff
@@ -124,26 +130,41 @@ contract DxMgnPool is Ownable {
+        currentState = State.DepositWithdrawnFromDx;
     }
 
+    bool public isMagnoliaWithdrawnFromDX = false;
```
Is this still needed? Once we have executed this method, we will be in `State.MgnUnlocked` and thus should not be able to call this method again.

josojo replied:
thanks for notifying me. Yes, it is not needed at all..

fleupold commented on contracts/DxMgnPool.sol:
```diff
@@ -56,6 +58,9 @@ contract DxMgnPool is Ownable {
      * Public interface
      */
     function deposit(uint amount) public {
+        checkForStateUpdate();
```
It's a little bit dangerous, that we have to remember to call this method now on every method that relies to have an up to date state. It was nicer, that it was implicitly calculated before, and didn't require a write to a local variable.

The alternative would be to still require access to the variable via a method (`currentState()`) and do the state update implicitly in there. But I'm not sure implicit state mutation in a getter is a good idea either.

Let's leave it like that.

josojo replied:
Yeah, I agree with your assessment

fleupold commented on contracts/DxMgnPool.sol:
```diff
@@ -124,26 +130,38 @@ contract DxMgnPool is Ownable {
+        // require(unlockedTokens[msg.sender].withdrawalTime < now, "The tokens cannot be withdrawn yet");
+
         totalMgn = mgnToken.balanceOf(address(this));
+        mgnToken.withdrawUnlockedTokens();
```
I'm not sure you can invert the order here. As far as I understand the FRTToken contract, we don't add tokens to the balance before `withdrawUnlockedTokens`, thus `balanceOf` would likely be 0 before.

josojo replied:
Uh this slipped in. Thanks a lot for the careful review!

Why do my tests still pass? WTF

fleupold replied:
The e2e tests should have caught this (if it's not already the case, can we make sure to add one that would). Unfortunately such an issue is not caught by the unit tests because of mocking.

fleupold commented on test/dx_mgn_pool.js:
```diff
@@ -400,17 +460,39 @@ contract("DxMgnPool", (accounts) => {
 
       await truffleAssert.reverts(instance.triggerMGNunlockAndClaimTokens(), "Pooling period is not yet over.")
     })
+    it("can deal with the case, totalDeposit == 0 (e.g. all funds are lost in dutchX)", async () => {
```
"can deal with the case" is not a really descriptive behaviour. Can you clarify what the expected behaviour is? Also "funds are lost" sounds a bit like we are hacked. Maybe "balance of deposits is 0"?

josojo commented:
I have fixed the finding and added the tests.

Although I really don't think the idempotent tests are needed with such a state machine. We anyways check that we are going into all states and the only transformation can happen in the triggered functions.

Anyways, safe is safe, tests are there.

fleupold commented on test/dx_mgn_pool.js:
```diff
@@ -443,31 +549,68 @@ contract("DxMgnPool", (accounts) => {
-      await truffleAssert.reverts(instance.withdrawUnlockedMagnoliaFromDx(), "Pooling period is not yet over.")
+      await truffleAssert.reverts(instance.withdrawUnlockedMagnoliaFromDx(), "Unlocking not yet triggered")
+    })
+    it("Can not be called twice (idempotenz)", async () => {
```
```suggestion
    it("can not be called twice (idempotence)", async () => {
```

fleupold commented on test/dx_mgn_pool.js:
```diff
@@ -400,17 +460,63 @@ contract("DxMgnPool", (accounts) => {
+      
+      instance.triggerMGNunlockAndClaimTokens()
+    })
+    it("checks idempotenz of funciton", async () => {
```
```suggestion
    it("can not be called twice (idempotence)", async () => {
```

fleupold commented on test/dx_mgn_pool.js:
```diff
@@ -400,17 +460,63 @@ contract("DxMgnPool", (accounts) => {
 
       await truffleAssert.reverts(instance.triggerMGNunlockAndClaimTokens(), "Pooling period is not yet over.")
     })
+    it("can deal with the case, totalDeposit == 0 (e.g. all funds are lost in dutchX)", async () => {
```
```suggestion
    it("allows claiming MGN if value of deposits in dutchX is 0", async () => {
```
